### PR TITLE
UFW module (another take)

### PR DESCRIPTION
**Pull request by jkeskikangas**
Based on @ovcharenko #5128. Additions
- Separate `state`, `policy` and `rule` attributes
- Support for `logging` command
- Support for `direction` and `interface` attributes
- Simple and robust change notifications based on `ufw status verbose` diff

Feedback welcome.

### CODE REVIEW

**Comment by ovcharenko** [comment]
Good job! But I think you need create separate pull request containing ufw module only. I merged your changes... so I'm fine with whatever pull request they will select for merge.

**Comment by jkeskikangas** [comment]
Thanks @ovcharenko! Whoa, I don't know how I messed up the pull request, it was fine a while ago. I'll see if I can fix it tomorrow. Luckily at least your's is fine.

-- edit
I fixed the pull request to contain only the relevant commits.

**Comment by jkeskikangas** [comment]
I cleaned up the implementation and updated the docs. Ready for review.

**Comment by ovcharenko** [comment]
Looks great!

File: library/system/ufw [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    - See C(man ufw) for more examples.
+requirements:
+    - C(ufw) package
+options:
+  state:
+    description:
+      - C(enabled) reloads firewall and enables firewall on boot.
+      - C(disabled) unloads firewall and disables firewall on boot.
+      - C(reloaded) reloads firewall.
+      - C(reseted) disables and resets firewall to installation defaults.
```
Comment (dbrgn) [inline library/system/ufw:41]: Is `reseted` proper english? Shouldn't it be just `reset`? (I'm not a native speaker so I'm unsure about that :))

File: library/system/ufw [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    default: 'any'
+  to_port:
+    description:
+      - Destination port.
+    required: false
+    aliases: ['port']
+  proto:
+    description:
+      - TCP/IP protocol.
+    choices: ['any', 'tcp', 'udp', 'ipv6']
```
Comment (dbrgn) [inline library/system/ufw]: What about support for `esp` and `ah` protocols (IPSec)? :) They seem to be supported according to the man page:

```
IPSec  is supported by using the 'esp' ('50') and 'ah' ('51') protocols. These protocols can only
be used with the full syntax. For example:

  ufw allow to 10.0.0.1 proto esp
  ufw allow to 10.0.0.1 from 10.4.0.0/16 proto esp
  ufw allow to 10.0.0.1 proto ah
  ufw allow to 10.0.0.1 from 10.4.0.0/16 proto ah
```

**Comment by dbrgn** [comment]
Something I noticed: When deploying some rules onto a new server, ansible did not report the tasks as "changed" even though the firewall rules were newly created (and work). Might that be due to the firewall being disabled when adding the rules?

Otherwise the module seems to work nicely. Thanks for creating it.

File: library/system/ufw (hunk 1 above)
Comment (jkeskikangas) [inline library/system/ufw:41]: Thanks, that's correct. I'll fix the typo.

File: library/system/ufw (hunk 2 above)
Comment (jkeskikangas) [inline library/system/ufw]: Thanks, that's true. I'll add the options.

Comment (artob) [inline library/system/ufw]: The list of supported protocol names surely is sourced from `/etc/protocols` and/or `/etc/services`, so adding support for protocol names one by one might not be a tenable approach?

**Comment by anderiv** [comment]
This looks great! Has it been targeted to a release yet?

I was hoping it would make it into 1.5, but doesn't look like that happened.

File: library/system/ufw (hunk 1 above)
Comment (jkeskikangas) [inline library/system/ufw:41]: Fixed in 0707462.

File: library/system/ufw (hunk 2 above)
Comment (jkeskikangas) [inline library/system/ufw]: Thanks @dbrgn, support for `esp` and `ah` added in ca2ab4c. 

@bendiken iptables (and ufw) works on transport layer (tcp, udp) and Internet layer (ip, ipsec) protocols. /etc/services actually maps well-known service names to tcp/udp ports, as explained in the [man page](http://manpages.ubuntu.com/manpages/precise/man5/services.5.html).

**Comment by jkeskikangas** [comment]
@dbrgn, the ansible 'changed' notifications for the ufw module are based on diffing the output of `ufw status verbose` before and after running the actual command.

If the firewall is disabled as new rules are added, they are not regarded as changing the (running) state of the firewall. Please see the following outputs run on two clean slate ubuntu 12.04:

**Firewall enabled as a first step**

```
TASK: [base | Enable firewall] ************************************************
changed: [foo]

TASK: [base | Allow tcp/22 for SSH] *******************************************
changed: [foo]

TASK: [base | Reject other ports] *********************************************
changed: [foo]
```

**Firewall enabled as a last step**

```
TASK: [base | Allow tcp/22 for SSH] *******************************************
ok: [bar]

TASK: [base | Reject other ports] *********************************************
ok: [bar]

TASK: [base | Enable firewall] ************************************************
changed: [bar]
```

**Comment by dbrgn** [comment]
Can we maybe fix that behavior? I think a change should be reflected (as the rules _are_ changed, they're just not enabled.)

**Comment by jkeskikangas** [comment]
It seems that diffing `/lib/ufw/user*.rules` in addition to `ufw status` might work: http://askubuntu.com/questions/30781/see-configured-rules-even-when-inactive

I'll add an additional check by running 

```
grep '^### tuple' /lib/ufw/user*.rules
```

before and after executing the ufw command.

**Comment by jkeskikangas** [comment]
@dbrgn rule changes are now recognized properly even if ufw is in disabled state. Implemented in 37ef39d.

**Comment by jimi-c** [comment]
Merged, thanks!

**Pull request merged**